### PR TITLE
[22.05] Add CASTEP datatypes

### PR DESCRIPTION
**Pull request by patrick-austin**
NB: Didn't know what the "latest" release branch was for the purposes of adding datatypes, so have created another PR against 22.01 as well: #14485 

Added datatypes for a selection of CASTEP input and output files, which are used in the context of muon spectroscopy:
- `.castep`
- `.param`
- `.den_fmt`

See:
https://www.tcm.phy.cam.ac.uk/castep/documentation/WebHelp/content/modules/castep/expcastepfileformats.htm
https://www.tcm.phy.cam.ac.uk/castep/documentation/WebHelp/content/modules/castep/keywords/k_write_form_dens_castep.htm
https://github.com/muon-spectroscopy-computational-project/muon-galaxy/issues/15

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

**Comment by elichad** [comment]
If it wasn't already clear, @patrick-austin is doing some work with us in the muon team at STFC :) These file types will be useful for our tools.

**Comment by nsoranzo** [comment]
Welcome @patrick-austin ! 

> NB: Didn't know what the "latest" release branch was for the purposes of adding datatypes, so have created another PR against 22.01 as well

Currently 22.01 is the latest "tagged" release branch, so probably this PR won't be necessary since #14485 will be merged forward once accepted, but let's keep this open in case 22.05 is released first.

**Comment by mvdbeek** [comment]
In the meantime we've release 22.05, so we should merge this one and close #14485

File: lib/galaxy/datatypes/text.py [hunk 1]
```diff
@@ -1091,3 +1091,116 @@ def _looks_like_yaml(self, file_prefix: FilePrefix):
...
+        """
+        begin_header = "BEGIN header"
+        end_header = 'END header: data is "<a b c> charge" in units of electrons/grid_point * number'
+        grid_points = "of grid_points"
+        handle = file_prefix.string_io()
+        lines = handle.readlines()
+        if lines[0].strip() == begin_header and lines[9].strip() == end_header and lines[10].strip() == grid_points:
+            return True
+        else:
+            return False
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/text.py]:
```suggestion
        return lines[0].strip() == begin_header and lines[9].strip() == end_header and lines[10].strip() == grid_points
```

File: lib/galaxy/datatypes/text.py [hunk 2]
```diff
@@ -1091,3 +1091,116 @@ def _looks_like_yaml(self, file_prefix: FilePrefix):
...
+
+        # check the TASK keyword is present
+        # and that it is set to a valid CASTEP task
+        pattern = re.compile(r"^TASK ?: ?([A-Z\+]*)$", flags=re.IGNORECASE | re.MULTILINE)
+        task = file_prefix.search(pattern)
+        if task and task.group(1).upper() in valid_tasks:
+            return True
+        else:
+            return False
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/text.py]:
```suggestion
        return task and task.group(1).upper() in valid_tasks
```

**Review by mvdbeek — COMMENTED** [review]
Looks great to me. Any chance you could reduce the size of Si.den_fmt ?

**Review by mvdbeek — APPROVED** [review]
Is it OK if I rebase this into a single commit to get the larger test data out of the git history ?

**Comment by elichad** [comment]
Yes, go ahead :)

**Comment by mvdbeek** [comment]
Looks like I don't have the permissions to do that

**Comment by jmchilton** [comment]
Merged as https://github.com/galaxyproject/galaxy/pull/14504 instead, thanks for the contribution!

**Pull request closed without merging**